dali/pipeline/operators/decoder/nvjpeg_decoder_new.h

```diff
@@ -0,0 +1,315 @@
+    output.set_type(type);
+    output.Resize(output_shape_);
+    output.SetLayout(DALI_NHWC);
+
+    for (int i = 0; i < batch_size_; i++) {
+      const EncodedImageInfo& info = output_info_[i];
+      auto *output_data = output.mutable_tensor<uint8_t>(i);
+
+      if (info.nvjpeg_support) {
+        nvjpegImage_t nvjpeg_image;
+        nvjpeg_image.channel[0] = output_data;
+        nvjpeg_image.pitch[0] = NumberOfChannels(output_image_type_) * info.widths[0];
+
+        NVJPEG_CALL(nvjpegStateAttachDeviceBuffer(image_states_[i], device_buffer_));
+
+        NVJPEG_CALL(nvjpegDecodeJpegTransferToDevice(
+            handle_,
+            image_decoders_[i],
+            image_states_[i],
+            jpeg_streams_[i],
+            ws->stream()));
+
+        NVJPEG_CALL(nvjpegDecodeJpegDevice(
+            handle_,
+            image_decoders_[i],
+            image_states_[i],
+            &nvjpeg_image,
+            ws->stream()));
+      } else {
+        auto& in = ws->Input<CPUBackend>(0, i);
+        const auto* input_data = in.data<uint8_t>();
+        const auto& file_name = in.GetSourceInfo();
+        OCVFallback(input_data, in.size(), output_data, ws->stream(), file_name);
+      }
+    }
+  }
+
+  bool ShouldBeHybrid(EncodedImageInfo& info) {
+    return info.widths[0] * info.heights[0] > hybrid_huffman_threshold_;
+  }
+  /**
+   // Running the CPU stage
+   * Fallback to openCV's cv::imdecode for all images nvjpeg can't handle
+   */
+  void OCVFallback(const uint8_t *data, int size,
+                   uint8_t *decoded_device_data, cudaStream_t s, string file_name) {
+    const int c = (output_image_type_ == DALI_GRAY) ? 1 : 3;
+    auto decode_type = (output_image_type_ == DALI_GRAY) ? cv::IMREAD_GRAYSCALE
+                                                         : cv::IMREAD_COLOR;
+    cv::Mat input(1,
+                  size,
+                  CV_8UC1,
+                  reinterpret_cast<unsigned char*>(const_cast<uint8_t*>(data)));
+    cv::Mat tmp = cv::imdecode(input, decode_type);
+
+    if (tmp.data == nullptr) {
+      DALI_FAIL("Unsupported image type: " + file_name);
+    }
+
+    // Transpose BGR -> output_image_type_ if needed
+    if (IsColor(output_image_type_) && output_image_type_ != DALI_BGR) {
+      OpenCvColorConversion(DALI_BGR, tmp, output_image_type_, tmp);
+    }
+
+    CUDA_CALL(cudaMemcpyAsync(decoded_device_data,
+                              tmp.ptr(),
+                              tmp.rows * tmp.cols * c,
+                              cudaMemcpyHostToDevice, s));
+  }
+
+  USE_OPERATOR_MEMBERS();
+  nvjpegHandle_t handle_;
+
+  // output colour format
+  DALIImageType output_image_type_;
+
+  unsigned hybrid_huffman_threshold_;
+
+  // Common
+  // Storage for per-image info
+  std::vector<EncodedImageInfo> output_info_;
+  nvjpegJpegDecoder_t decoder_huff_host_;
+  nvjpegJpegDecoder_t decoder_huff_hybrid_;
+
+  // CPU
+  std::vector<nvjpegJpegDecoder_t> image_decoders_;
```

Comment:
Just an idea:
I am thinking that it'd be cleaner to have one vector of structs containing all those nvjpeg* types instead of N vectors

Reply:
Sure, could reuse what I did for the splitted version